### ISSUE TITLE
Registry class needs generic 'apply' method (Enhancement)

### ISSUE THREAD

**Issue by meldavis** [gabime/spdlog#238]
Please add this or similar function to the registry class:
```c++
void apply(std::function<void(std::shared_ptr<logger>)> functor) {
    std::lock_guard<Mutex> lock(_mutex);
	for (auto& l : _loggers)
	    functor(l.second);
}
```
## Purpose ##
The *logger::_sink_it* method is virtual so that custom loggers can derive from the logger class.  It's likely that a derived logger will have added properties that should be set globally, similar to setting the *level* on spdlog::loggers.     Presently, there is no facility to set  custom logger properties under the protection of a mutex, other than creating a custom *spdlog::details::registry* class.     Adding this one method will address that. 

Here is an example of its use.  Assume that the program has a log::manager class.
```c++
void manager::set_fatal_policy(level lvl, fatal_action a) {
    spdlog::details::registry::instance().apply(
        [lvl, a](auto& l) { 
            auto lg = std::dynamic_pointer_cast<custom_logger>(l);
            lg->set_fatal_level(lvl);
            lg->set_fatal_action(a);
         }
    );
}
```

**Comment by gabime** [comment]
I am not sure i understand. Why the sdlog::registry class need to be involved?

**Comment by meldavis** [comment]
For the same reasons, I suppose, that the spdlog::registry has a 
```c++
void set_level(level::level_enum log_level)
```


At some point in execution, the program needs to reset a property so that all loggers have the same value.  The registry holds the iterator to all the loggers.   Just as the *set_level* method iterates over all the the loggers to set them to the same log_level; the *apply* method would iterate over all the registered loggers to that the manager class can apply a common setting. 

Without this method, the program needs to maintain a separate list loggers and a mutex so that no logger is added or removed while the setting is being changed.  Effectively, it would have to duplicate the registry.

**Comment by gabime** [comment]
Oh Ok , so the intention is to provide a way to iterate over all the registered loggers in a thread safe way ?

Maybe visit_all would be better name:

```c++
spdlog::visit_all([](auto& logger) { /* Do something with the logger */});
```

**Comment by meldavis** [comment]
Sounds great.

**Comment by gabime** [comment]
Implemented in 3af247fbd3b23b393197b3ac0afbd3f9ac0781f3
